File: src/query/Stream.php
<?php

namespace Ething\Query;



class Stream {
	
	private $index = 0;
	private $index_last = 0;
	private $content = '';
	
	public function __construct($content) {
		if(is_string($content)){
			$this->content = $content;
		}
	}
	
	public function currentIndex(){
		return $this->index;
	}
	
	public function previousIndex(){
		return $this->index_last;
	}
	
	public function walk($numberOfChar){
		if($numberOfChar>0){
			$length = strlen($this->content);
			if($numberOfChar<$length){
				$this->content = substr($this->content,$numberOfChar);
			}
			else {
				$numberOfChar = $length;
				$this->content = '';
			}
		}
		$this->index_last = $this->index;
		$this->index += $numberOfChar;
		return $numberOfChar;
	}
	
	public function read($a){
		if(is_int($a) && $a>0){
			// read n characters
			$o = substr($this->content, 0, $a);
			$this->walk(strlen($o));
			return $o;
		}
		else if(is_string($a)){
			// return the first match
			if(preg_match($a,$this->content,$match,PREG_OFFSET_CAPTURE)){
				$o = $match[0][0];
				$this->walk(strlen($o)+$match[0][1]);
				return $o;
			}
			else
				$this->walk(0);// just for updating the $index_last property
		}
		return null;
	}
	
	public function match($regex){
		return (bool) preg_match($regex,$this->content);
	}
	
	public function skipSpace(){
		return $this->read('/^\s*/');
	}
	
	public function readWord(){
		return $this->read('/[^\s]+/');
	}
	
	public function length(){
		return strlen($this->content);
	}
	
	public function __toString(){
		return $this->content;
	}
}

File: ething/query/Stream.py
# coding: utf-8
from future.utils import string_types, integer_types


import re


class Stream(object):

    def __init__(self, content=None):
        self.index = 0
        self.index_last = 0
        if isinstance(content, string_types):
            self.content = content
        else:
            self.content = ''

    def currentIndex(self):
        return self.index

    def previousIndex(self):
        return self.index_last

    def walk(self, numberOfChar):
        if numberOfChar > 0:
            length = len(self.content)
            if numberOfChar < length:
                self.content = self.content[numberOfChar:]
            else:
                numberOfChar = length
                self.content = ''

        self.index_last = self.index
        self.index += numberOfChar
        return numberOfChar

    def read(self, a, ignore_case = False):
        if isinstance(a, integer_types) and a > 0:
            # read n characters
            o = self.content[:a]
            self.walk(len(o))
            return o

        elif isinstance(a, string_types):
            # return the first match
            m = re.match(a, self.content, flags = re.I if ignore_case else 0)
            if m is not None:
                o = m.group(0)
                self.walk(m.start(0)+len(o))
                return o
            else:
                self.walk(0)  # just for updating the index_last property

        return None

    def match(self, regex):
        m = re.match(regex, self.content)
        return bool(m)

    def skipSpace(self):
        return self.read('^\s*')

    def readWord(self):
        return self.read('[^\s]+')

    def length(self):
        return len(self.content)

    def __str__(self):
        return self.content
